Mathlib/Computability/NFA.lean

```diff
@@ -112,6 +120,275 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [← toDFA_correct] at hx ⊢
   exact M.toDFA.pumping_lemma hx hlen
 
+section Closure
+
+/-! ### NFA Closure Properties -/
+
+section Reversal
+
+section Auxilary
+
+/-- `M.unstep` is the reverse of `M.step`. -/
+def unstep (s : σ) (a : α) : Set σ := {s' | s ∈ M.step s' a}
+
+theorem mem_unstep {s t : σ} {a : α} : s ∈ M.unstep t a ↔ t ∈ M.step s a := Iff.rfl
+
+/-- Reversed analog of `M.stepSet S a`:
+  `M.unstepSet S a` is the union of `unstep M s a` for all `s ∈ S`.
+  It computes the set of states that have a transition in `M`
+  to a state in `S`. -/
+def unstepSet (S : Set σ) (a : α) : Set σ := ⋃ s ∈ S, M.unstep s a
+
+theorem mem_unstepSet {s : σ} {S : Set σ} {a : α} :
+    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, s ∈ M.unstep t a := by
+  simp [unstepSet]
+
+theorem mem_unstepSet_step {s : σ} {S : Set σ} {a : α} :
+    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, t ∈ M.step s a := by
+  simp [mem_unstepSet, mem_unstep]
+
+/-- Reseversed analog of `M.evalFrom S x`:
+  `M.rewindFrom S x` computes all possible reversed paths through `M` with
+  input `x` starting at an element of `S`. -/
+def rewindFrom : Set σ → List α → Set σ := List.foldl M.unstepSet
+
+/-- `M.rewind x` computes all possible paths through `M` with input `x` ending at an element of
+  `M.accept`. -/
+def rewind : List α → Set σ := M.rewindFrom M.accept
+
+/-- `M.rewindsToStart S` is the language of `x`
+  such that starting from `S` we rewind to `M.start`. -/
+def rewindsToStart (S : Set σ) : Language α :=
+  { x | ∃ s ∈ M.start, s ∈ M.rewindFrom S x }
+
+lemma mem_rewindsToStart {S : Set σ} {x : List α} :
+    x ∈ M.rewindsToStart S ↔ ∃ s ∈ M.start, s ∈ M.rewindFrom S x := Iff.rfl
+
+end Auxilary
+
+/-- NFAs are closed under reversal:
+  Given NFA `M`, there is an NFA `reverse(M)` such that
+  `L(reverse(M)) = reverse(L(M))`. -/
+def reverse (M : NFA α σ) : NFA α σ where
+  step := M.unstep
+  start := M.accept
+  accept := M.start
+
+lemma reverse_acceptsFrom_rewindsToStart {M : NFA α σ} :
+    M.reverse.acceptsFrom = M.rewindsToStart := by
+  ext xs
+  rfl
+
+lemma reverse_rewindFrom_evalFrom {xs : List α} {S1 S2 : Set σ} {M : NFA α σ} :
+    (∃ s1 ∈ S1, s1 ∈ M.rewindFrom S2 xs) ↔
+    (∃ s2 ∈ S2, s2 ∈ M.evalFrom S1 xs.reverse) := by
+  dsimp [evalFrom, rewindFrom]
+  rw [List.foldl_reverse]
+  revert S1 S2
+  induction xs <;> intros S1 S2 <;> simp
+  case nil =>
+    constructor <;> rintro ⟨s, h1, h2⟩ <;> exists s
+  case cons x xs ih =>
+    rw [ih]
+    clear ih
+    unfold unstepSet
+    unfold stepSet
+    constructor
+    · rintro ⟨s3, h, h1⟩
+      simp [Set.mem_iUnion] at h
+      rcases h with ⟨s2,h2,h3⟩
+      exists s2
+      constructor <;> try assumption
+      simp [Set.mem_iUnion]
+      exists s3
+    · rintro ⟨s2, h2, h⟩
+      simp [Set.mem_iUnion] at h
+      rcases h with ⟨s3, h1, h3⟩
+      exists s3
+      simp [Set.mem_iUnion]
+      constructor <;> try assumption
+      exists s2
```

Comment:
Golfed:

```lean
lemma reverse_rewindFrom_evalFrom {xs : List α} {S1 S2 : Set σ} {M : NFA α σ} :
    (∃ s1 ∈ S1, s1 ∈ M.rewindFrom S2 xs) ↔
    (∃ s2 ∈ S2, s2 ∈ M.evalFrom S1 xs.reverse) := by
  dsimp [evalFrom, rewindFrom]
  rw [List.foldl_reverse]
  induction xs generalizing S1 S2
  case nil => tauto
  case cons x xs ih =>
    simp [ih, unstepSet, stepSet, Set.mem_iUnion]
    tauto
```

Of note:
- `revert; induction; intros` is just `induction ... generalizing`
- I can't find it documented right now, but I believe it is not good practice to have [unsqueezed `simp` calls in non-terminal positions](https://leanprover-community.github.io/contribute/style.html#squeezing-simp-calls) because such proofs break easily.

Reply:
Also, the statement is quite weird. "exists s1 in S1 such that s1 is also in blah" is simply "S1 is a subset of blah". I think that's how you want to do this? Or maybe prove a more general `S₁ ⊆ S₂ -> M.evalFrom S₁ x ⊆ M.evalFrom S₂ x` and use that to simplify things.

Reply:
Thanks for the feedback! I will try the subset style definition.

Reply:
>I can't find it documented right now, but I believe it is not good practice to have [unsqueezed simp calls in non-terminal positions](https://leanprover-community.github.io/contribute/style.html#squeezing-simp-calls) because such proofs break easily.

What is the general fix for this?

Reply:
> What is the general fix for this?

I found [some more relevant documentation](https://leanprover-community.github.io/extras/simp.html#non-terminal-simps) after I wrote that comment, perhaps that'll be helpful.

Reply:
> Thanks for the feedback! I will try the subset style definition.

oh, sorry, looks like I've misled you, I had a brainfart. what I'm saying would be true if it was a forall. since it's an exists, you could state it as "the intersection of these two sets is nonempty". not sure which would work better, but writing this as an intersection seems like it would be more flexible (because intersection is easily commutative). your call.

Reply:
> > Thanks for the feedback! I will try the subset style definition.
> 
> oh, sorry, looks like I've misled you, I had a brainfart. what I'm saying would be true if it was a forall. since it's an exists, you could state it as "the intersection of these two sets is nonempty". not sure which would work better, but writing this as an intersection seems like it would be more flexible (because intersection is easily commutative). your call.

What do you mean? Is it ok to leave as is?